### ISSUE TITLE
WaveNet

### ISSUE THREAD

**Issue by JYKlabs** [JYKlabs/JYKlabs#3]
import os
import torch
import torch.nn as nn
import torch.optim as optim
import torchaudio
from torch.utils.data import Dataset, DataLoader, SubsetRandomSampler
from sklearn.model_selection import train_test_split
import time
from openpyxl import Workbook
import shutil

def print_log(message):
    print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {message}")

class ConditionalWaveNet(nn.Module):
    def __init__(self, num_channels=1, residual_channels=32, dilation_channels=32, skip_channels=256, 
                 kernel_size=2, num_layers=10, gain_values=7, peak_values=11):
        super(ConditionalWaveNet, self).__init__()
        self.num_channels = num_channels
        self.mu = 256
        self.embedding = nn.Embedding(self.mu, residual_channels)
        self.gain_embedding = nn.Embedding(gain_values, residual_channels)
        self.peak_embedding = nn.Embedding(peak_values, residual_channels)
        self.dilations = [2**i for i in range(num_layers)] * 2
        self.convs = nn.ModuleList()
        self.skips = nn.ModuleList()
        for dilation in self.dilations:
            self.convs.append(nn.Conv1d(residual_channels, dilation_channels, kernel_size, dilation=dilation, padding=0))
            self.skips.append(nn.Conv1d(dilation_channels, skip_channels, kernel_size=1))
        self.post_conv1 = nn.Conv1d(skip_channels, skip_channels, kernel_size=1)
        self.post_conv2 = nn.Conv1d(skip_channels, self.mu, kernel_size=1)
        self.relu = nn.ReLU()
        self.softmax = nn.Softmax(dim=1)

    def forward(self, x, gain, peak):
        x = self.embedding(x).transpose(1, 2)
        gain_embed = self.gain_embedding(gain).unsqueeze(2)
        peak_embed = self.peak_embedding(peak).unsqueeze(2)
        condition = gain_embed + peak_embed
        skip_connections = []
        for conv, skip in zip(self.convs, self.skips):
            pad = (conv.kernel_size[0] - 1) * conv.dilation[0]
            x_padded = nn.functional.pad(x, (pad, 0))
            x = self.relu(conv(x_padded) + condition)
            skip_out = skip(x)
            skip_connections.append(skip_out)
        out = sum(skip_connections)
        out = self.relu(self.post_conv1(out))
        out = self.post_conv2(out)
        out = self.softmax(out)
        return out

class WaveNetDataset(Dataset):
    def __init__(self, base_dir, input_prefix, sample_rate=16000, seq_len=16000):
        self.base_dir = base_dir
        self.input_prefix = input_prefix
        self.sample_rate = sample_rate
        self.seq_len = seq_len
        self.input_files = []
        self.output_files = []
        self.gain_indices = []
        self.peak_indices = []
        self.gain_map = {5: 0, 10: 1, 15: 2, 20: 3, 25: 4, 30: 5, 35: 6}
        self.peak_map = {i: idx for idx, i in enumerate(range(0, 101, 10))}

        for peak in range(0, 101, 10):
            folder = os.path.join(base_dir, f"Peak Reduction {peak}")
            if not os.path.exists(folder):
                continue
            input_files = [f for f in os.listdir(folder) if f.startswith(input_prefix) and f.endswith(".wav")]
            for f in input_files:
                self.input_files.append(os.path.join(folder, f))
                self.output_files.append(os.path.join(folder, f.replace(input_prefix, "outboardinput_")))
                parts = f.split("_")
                gain = int([p for p in parts if p.startswith("gain")][0].replace("gain", ""))
                peak_val = int([p for p in parts if p.startswith("peak")][0].replace("peak", ""))
                self.gain_indices.append(self.gain_map[gain])
                self.peak_indices.append(self.peak_map[peak_val])
        
        print_log(f"Found {len(self.input_files)} files with prefix {input_prefix}")

    def __len__(self):
        return len(self.input_files)

    def mu_law_encode(self, audio, mu=255):
        audio = torch.clamp(audio, -1, 1)
        return torch.sign(audio) * torch.log1p(mu * torch.abs(audio)) / torch.log1p(torch.tensor(mu))

    def __getitem__(self, idx):
        input_path = self.input_files[idx]
        output_path = self.output_files[idx]
        input_wav, sr = torchaudio.load(input_path)
        output_wav, _ = torchaudio.load(output_path)
        if sr != self.sample_rate:
            input_wav = torchaudio.transforms.Resample(sr, self.sample_rate)(input_wav)
            output_wav = torchaudio.transforms.Resample(sr, self.sample_rate)(output_wav)
        if input_wav.shape[1] > self.seq_len:
            start = torch.randint(0, input_wav.shape[1] - self.seq_len, (1,)).item()
            input_wav = input_wav[:, start:start + self.seq_len]
            output_wav = output_wav[:, start:start + self.seq_len]
        else:
            pad_len = self.seq_len - input_wav.shape[1]
            input_wav = nn.functional.pad(input_wav, (0, pad_len))
            output_wav = nn.functional.pad(output_wav, (0, pad_len))
        input_quant = self.mu_law_encode(input_wav).mul(128).add(128).long().squeeze(0)
        output_quant = self.mu_law_encode(output_wav).mul(128).add(128).long().squeeze(0)
        return input_quant, output_quant, self.gain_indices[idx], self.peak_indices[idx]

def split_dataset(dataset, train_ratio=0.8, val_ratio=0.1):
    dataset_size = len(dataset)
    indices = list(range(dataset_size))
    train_idx, temp_idx = train_test_split(indices, train_size=train_ratio, random_state=42)
    val_idx, test_idx = train_test_split(temp_idx, train_size=val_ratio/(val_ratio + 0.1), random_state=42)
    return train_idx, val_idx, test_idx

print_log("Testing log function before train_wavenet")
def train_wavenet(model, train_loader, val_loader, test_loader, device, num_epochs=30, model_name="wavenet", save_dir="/Volumes/Mac Hub 3 for Dropbox/wavenet_save"):
    global print_log
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    model_dir = os.path.join(save_dir, model_name)
    os.makedirs(model_dir, exist_ok=True)
    print_log(f"Created model directory: {model_dir}")
    
    start_time = time.time()
    print_log(f"Starting training for {model_name}")
    
    epoch_logs = []
    batch_logs = {f"Epoch {i+1}": [] for i in range(num_epochs)}
    total_batches = num_epochs * len(train_loader)
    batch_count = 0

    for epoch in range(num_epochs):
        model.train()
        train_loss = 0.0
        epoch_start = time.time()
        batch_losses = []
        print_log(f"Starting Epoch {epoch+1}/{num_epochs}")
        for i, (inputs, targets, gains, peaks) in enumerate(train_loader):
            batch_count += 1
            progress = (batch_count / total_batches) * 100
            inputs, targets = inputs.to(device), targets.to(device)
            gains, peaks = gains.to(device), peaks.to(device)
            optimizer.zero_grad()
            outputs = model(inputs, gains, peaks)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
            train_loss += loss.item()
            batch_losses.append(loss.item())
            print(f"{model_name} - Epoch [{epoch+1}/{num_epochs}], Batch [{i+1}/{len(train_loader)}], "
                  f"Loss: {loss.item():.4f}, Progress: {progress:.2f}%", end='\r')
        
        train_loss /= len(train_loader)
        batch_logs[f"Epoch {epoch+1}"] = batch_losses

        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for inputs, targets, gains, peaks in val_loader:
                inputs, targets = inputs.to(device), targets.to(device)
                gains, peaks = gains.to(device), peaks.to(device)
                outputs = model(inputs, gains, peaks)
                val_loss += criterion(outputs, targets).item()
        val_loss /= len(val_loader)

        epoch_log = f"Epoch [{epoch+1}/{num_epochs}], Train Loss: {train_loss:.4f}, Val Loss: {val_loss:.4f}, Time: {(time.time() - epoch_start) // 60:.0f}m {(time.time() - epoch_start) % 60:.2f}s"
        epoch_logs.append(epoch_log)
        print(f"\n{model_name} - {epoch_log}")
        
        torch.save(model.state_dict(), os.path.join(model_dir, f"{model_name}_epoch_{epoch+1}.pth"))
        print_log(f"Saved model: {model_dir}/{model_name}_epoch_{epoch+1}.pth")
        time.sleep(1)

    model.eval()
    test_loss = 0.0
    with torch.no_grad():
        for inputs, targets, gains, peaks in test_loader:
            inputs, targets = inputs.to(device), targets.to(device)
            gains, peaks = gains.to(device), peaks.to(device)
            outputs = model(inputs, gains, peaks)
            test_loss += criterion(outputs, targets).item()
    test_loss /= len(test_loader)
    print_log(f"{model_name} - Test Loss: {test_loss:.4f}")

    excel_file = os.path.join(save_dir, f"{model_name}_training_logs.xlsx")
    backup1_file = os.path.join(save_dir, f"{model_name}_training_logs_backup1.xlsx")
    backup2_file = os.path.join(save_dir, f"{model_name}_training_logs_backup2.xlsx")
    print_log(f"Saving training logs to {excel_file}")
    wb = Workbook()
    ws_epoch = wb.active
    ws_epoch.title = "Epoch Logs"
    ws_epoch.append(["Epoch Info"])
    for log_entry in epoch_logs:
        ws_epoch.append([log_entry])
    
    for epoch, losses in batch_logs.items():
        ws_batch = wb.create_sheet(title=epoch)
        ws_batch.append(["Batch", "Train Loss"])
        for i, loss in enumerate(losses, 1):
            ws_batch.append([i, loss])
    
    for attempt in range(5):
        try:
            wb.save(excel_file)
            shutil.copy(excel_file, backup1_file)
            shutil.copy(excel_file, backup2_file)
            if os.path.exists(excel_file) and os.path.exists(backup1_file) and os.path.exists(backup2_file):
                print_log(f"Successfully saved training logs to {excel_file}, backups to {backup1_file}, {backup2_file}")
                break
            else:
                raise FileNotFoundError("Files not created")
        except Exception as e:
            print_log(f"Attempt {attempt+1}/5 failed to save {excel_file}: {e}")
            time.sleep(2)
    else:
        raise RuntimeError(f"Failed to save {excel_file} after 5 attempts")

    total_time = time.time() - start_time
    print_log(f"Training finished, Total time: {total_time // 60:.0f}m {total_time % 60:.2f}s")

if __name__ == "__main__":
    base_dir = "/Volumes/Mac Hub 3 for Dropbox/Dropbox/연구/Audio Hardware Modeling/Data Set/Test Signals Results/Linear_chirp"
    save_dir = "/Volumes/Mac Hub 3 for Dropbox/wavenet_save"
    device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
    print_log(f"Using device: {device}")
    
    dataset_a = WaveNetDataset(base_dir=base_dir, input_prefix="Fixed_SA2A_input_")
    train_idx_a, val_idx_a, test_idx_a = split_dataset(dataset_a)
    train_loader_a = DataLoader(dataset_a, batch_size=6, sampler=SubsetRandomSampler(train_idx_a), num_workers=0, pin_memory=True)
    val_loader_a = DataLoader(dataset_a, batch_size=6, sampler=SubsetRandomSampler(val_idx_a), num_workers=0, pin_memory=True)
    test_loader_a = DataLoader(dataset_a, batch_size=6, sampler=SubsetRandomSampler(test_idx_a), num_workers=0, pin_memory=True)

    dataset_b = WaveNetDataset(base_dir=base_dir, input_prefix="directardinput_")
    train_idx_b, val_idx_b, test_idx_b = split_dataset(dataset_b)
    train_loader_b = DataLoader(dataset_b, batch_size=6, sampler=SubsetRandomSampler(train_idx_b), num_workers=0, pin_memory=True)
    val_loader_b = DataLoader(dataset_b, batch_size=6, sampler=SubsetRandomSampler(val_idx_b), num_workers=0, pin_memory=True)
    test_loader_b = DataLoader(dataset_b, batch_size=6, sampler=SubsetRandomSampler(test_idx_b), num_workers=0, pin_memory=True)
    
    print_log("Training WaveNet for Clone A...")
    model_a = ConditionalWaveNet(gain_values=7, peak_values=11).to(device)
    train_wavenet(model_a, train_loader_a, val_loader_a, test_loader_a, device, num_epochs=30, model_name="wavenet_clone_a", save_dir=save_dir)
    
    print_log("Training WaveNet for Clone B...")
    model_b = ConditionalWaveNet(gain_values=7, peak_values=11).to(device)
    train_wavenet(model_b, train_loader_b, val_loader_b, test_loader_b, device, num_epochs=30, model_name="wavenet_clone_b", save_dir=save_dir)
